### ISSUE TITLE
Inconsistent triggering of SC2268 (x-prefix comparisons)

### ISSUE THREAD

**Issue by slycordinator** [koalaman/shellcheck#2689]
SC2268 recommends that one not use x-prefix comparisons, but it allows them in many cases.

SC2268 appears to only get triggered if:
1) The comparison is = or ==. Nothing happens when the comparison is '!="
[ ! "x$var" = "x" ] and [ ! "x$var" == "x" ] trigger the warning, but [ "x$var" != "x" ] and [ ! "x$var" != "x" ] do not trigger it.

2) The prefix is solely the lowercase 'x'. Any other prefix doesn't trigger the issue. Recently came across scripts that occasionally have  "X" as the prefix and it doesn't get triggered. Also doesn't get triggered when any other letter is used.
[ "x$var" = "x" ] triggers this, but [ "X$var" = "X" ] does not.

Thank you

#### For bugs
- Rule Id 2068: 
- My shellcheck version (`shellcheck --version` or "online"): 
- [x] The rule's wiki page does not already cover this (e.g. https://shellcheck.net/wiki/SC2086)
- [X] I tried on https://www.shellcheck.net/ and verified that this is still a problem on the latest commit

#### For new checks and feature suggestions
- [x] https://www.shellcheck.net/ (i.e. the latest commit) currently gives no useful warnings about this
- [X] I searched through https://github.com/koalaman/shellcheck/issues and didn't find anything related

#### Here's a snippet or screenshot that shows the problem:

```sh

#!/bin/bash
#shellcheck disable=SC2154

if [ "x$var" = "x" ]; then
    whatever
fi

if [ "x$var" == "x" ]; then
    whatever
fi

if [ ! "x$var" == "x" ]; then
    whatever
fi

if [ "x$var" != "x" ]; then
    whatever
fi

if [ "X$var" != "X" ]; then
    whatever
fi

if [ "X$var" = "X" ]; then
    whatever
fi

if [ "X$var" == "X" ]; then
    whatever
fi

```

#### Here's what shellcheck currently says:
```
In whatever.sh line 4:
if [ "x$var" = "x" ]; then
     ^-----^ SC2268 (style): Avoid x-prefix in comparisons as it no longer serves a purpose.

Did you mean:
if [ "$var" = "" ]; then


In whatever.sh line 8:
if [ "x$var" == "x" ]; then
     ^-----^ SC2268 (style): Avoid x-prefix in comparisons as it no longer serves a purpose.

Did you mean:
if [ "$var" == "" ]; then


In whatever.sh line 12:
if [ ! "x$var" == "x" ]; then
       ^-----^ SC2268 (style): Avoid x-prefix in comparisons as it no longer serves a purpose.

Did you mean:
if [ ! "$var" == "" ]; then
```

#### Here's what I wanted or expected to see:
(in addition to the above)

```
In whatever.sh line 16:
if [ "x$var" != "x" ]; then
       ^-----^ SC2268 (style): Avoid x-prefix in comparisons as it no longer serves a purpose.

Did you mean:
if [ "$var" != "" ]; then

In whatever.sh line 20:
if [ "X$var" != "X" ]; then
       ^-----^ SC2268 (style): Avoid x-prefix in comparisons as it no longer serves a purpose.

Did you mean:
if [ "$var" != "" ]; then

In whatever.sh line 24:
if [ "X$var" = "X" ]; then
       ^-----^ SC2268 (style): Avoid x-prefix in comparisons as it no longer serves a purpose.

Did you mean:
if [ "$var" = "" ]; then

In whatever.sh line 24:
if [ "X$var" == "X" ]; then
       ^-----^ SC2268 (style): Avoid x-prefix in comparisons as it no longer serves a purpose.

Did you mean:
if [ "$var" == "" ]; then
```

**Comment by mcassaniti** [comment]
What is the preferred syntax here when $var could be undefined or the empty string? I ask because it is relevant for the documentation and one of the reasons I still use `[ "x$1" = "x--my-only-cmdline-option" ]` (yes I know I could use getopt).

**Comment by slycordinator** [comment]
Very old shells were confused when the first argument began with "-" as
well as when it was simply "!" or "(".

So, people would put x at the front of the argument/variable.

POSIX makes that unnecessary.

Unless you're on a shell that predates POSIX, you can replace
[ "x$1" = "x--my-only-cmdline-option" ]

with
[ "$1" = "--my-only-cmdline-option" ]
